nifi-nar-bundles/nifi-py4j-bundle/nifi-python-framework/src/main/python/framework/ExtensionManager.py

```diff
@@ -0,0 +1,531 @@
+            if (len(assignment_node.targets) == 1 and assignment_node.targets[0].id == 'version'):
+                assigned_values = assignment_node.value.value
+                logger.info("Found version of {0} to be {1}".format(class_name, assigned_values))
+                return assigned_values
+
+        # No dependencies found
+        logger.info("Found no version information for {0}".format(class_name))
+        return 'Unknown'
+
+
+    def __get_processor_dependencies__(self, details_node, class_name):
+        deps = self.__get_assigned_list__(details_node, class_name, 'dependencies')
+        if len(deps) == 0:
+            logger.info("Found no external dependencies that are required for class %s" % class_name)
+        else:
+            logger.info("Found the following external dependencies that are required for class {0}: {1}".format(class_name, deps))
+
+        return deps
+
+
+    def __get_processor_tags__(self, details_node, class_name):
+        return self.__get_assigned_list__(details_node, class_name, 'tags')
+
+
+    def __get_assigned_list__(self, details_node, class_name, element_name):
+        assignment_nodes = self.__get_assignment_nodes__(details_node)
+        for assignment_node in assignment_nodes:
+            if (len(assignment_node.targets) == 1 and assignment_node.targets[0].id == element_name):
+                assigned_values = assignment_node.value.elts
+                declared_dependencies = []
+                for assigned_value in assigned_values:
+                    declared_dependencies.append(assigned_value.value)
+
+                return declared_dependencies
+
+        # No values found
+        return []
+
+
+    def __get_processor_description__(self, details_node, class_name):
+        assignment_nodes = self.__get_assignment_nodes__(details_node)
+        for assignment_node in assignment_nodes:
+            if (len(assignment_node.targets) == 1 and assignment_node.targets[0].id == 'description'):
+                return assignment_node.value.value
+
+        # No description found
+        logger.debug("Found no description for class %s" % class_name)
+        return None
+
+
+
+    def __get_class_nodes__(self, node):
+        class_nodes = [n for n in node.body if isinstance(n, ast.ClassDef)]
+        return class_nodes
+
+
+    def __get_assignment_nodes__(self, node):
+        assignment_nodes = [n for n in node.body if isinstance(n, ast.Assign)]
+        return assignment_nodes
+
+
+    def __import_external_dependencies__(self, class_name, dependencies, target_dir):
+        if not os.path.exists(target_dir):
+            os.makedirs(target_dir)
+
+        completion_marker_file = os.path.join(target_dir, 'dependency-download.complete')
+        if os.path.exists(completion_marker_file):
+            logger.info("All dependencies have already been imported for {0}".format(class_name))
+            return True
+
+        success = False
+        if len(dependencies) > 0:
```

Comment:
If we are intent on providing a method to configure `pip` for the environment, I believe we should move away from the current single Python module approach. Unless we want to create our own solution to defining `pip.conf` with the current solution. I believe it would easier to allow providing `pip.conf`, `requirements.txt`, `pyproject.toml` and `setup.py` for an extension. Is there any technical reason we can't define the extensions as a standard Python package?